crates/matrix-sdk/src/room/joined.rs

```diff
@@ -621,29 +631,131 @@ impl Joined {
     ///         "My favorite cat",
     ///         &mime::IMAGE_JPEG,
     ///         &mut image,
-    ///         None,
+    ///         AttachmentConfig::new(),
     ///     ).await?;
     /// }
     /// # Result::<_, matrix_sdk::Error>::Ok(()) });
     /// ```
-    pub async fn send_attachment<R: Read>(
+    pub async fn send_attachment<'a, R: Read + Seek, T: Read>(
         &self,
         body: &str,
         content_type: &Mime,
         reader: &mut R,
-        txn_id: Option<&TransactionId>,
+        config: AttachmentConfig<'a, T>,
+    ) -> Result<send_message_event::Response> {
+        let reader = &mut BufReader::new(reader);
+
+        #[cfg(feature = "image_proc")]
+        let mut cursor;
+
+        if config.thumbnail.is_some() {
+            self.prepare_and_send_attachment(body, content_type, reader, config).await
+        } else {
+            #[cfg(not(feature = "image_proc"))]
+            let thumbnail = Thumbnail::NONE;
+
+            #[cfg(feature = "image_proc")]
+            let thumbnail = if config.generate_thumbnail {
+                match generate_image_thumbnail(content_type, reader, config.thumbnail_size) {
+                    Ok((thumbnail_data, thumbnail_info)) => {
+                        reader.rewind()?;
+
+                        cursor = Cursor::new(thumbnail_data);
+                        Some(Thumbnail {
+                            reader: &mut cursor,
+                            content_type: &mime::IMAGE_JPEG,
+                            info: Some(thumbnail_info),
+                        })
+                    }
+                    Err(error)
```

Comment:
Wouldn't it be nicer to match directly, i.e.:

```rust
match generate_image_thumbnail(content_type, reader, config.thumbnail_size) {
    Ok(...) => ...,
    Err(ImageError::ThumbnailBiggerThanOriginal) | (Err(ImageError::FormatNotSupported) => ...
    ...
}
```

This avoids the immediate double nesting.